flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -876,6 +887,17 @@ private void jobStatusChanged(
 		validateRunsInMainThread();
 
 		if (newJobStatus.isGloballyTerminalState()) {
+			// other terminal job states are handled by the executions
+			if (newJobStatus == JobStatus.FINISHED) {
+				scheduledExecutorService.execute(() -> {
+					for (Map.Entry<ResourceID, Tuple2<TaskManagerLocation, TaskExecutorGateway>> entry : registeredTaskManagers.entrySet()) {
```

Comment:
Do we have concurrent modification problem on the `registeredTaskManagers` ? Here we iterates the map in the scheduler executor, while main thread of JM will remove entries from this map in the method calls of `onStop` -> `disConnectTaskManger`.

Reply:
yes, we should run this in the JM main thread instead.